### ISSUE TITLE
Proposal: Allow adding interactive areas to the TitleBar

### ISSUE THREAD

**Issue by dotMorten** [microsoft/microsoft-ui-xaml#4233]
## Summary
Currently with Preview 4 we can extend into the titlebar with custom UI content - however any buttons, textboxes or others will not receive mouse inputs.

## Rationale
Having profile, search, and a few other quick-access features in the title bar are pretty common place now, but isn't possible with Preview 4.  While we can add info there now, none of it receives mouse input.

Here are some common scenarios where this is done today:

MS Edge Tabs:
![image](https://user-images.githubusercontent.com/1378165/108281408-8cd5ec00-7134-11eb-9bfa-377488e44e1f.png)

Mail App back button:
![image](https://user-images.githubusercontent.com/1378165/108281474-ae36d800-7134-11eb-8ec7-4e78c34fa0dd.png)

MS Paint:
![image](https://user-images.githubusercontent.com/1378165/108281550-cc043d00-7134-11eb-957c-d8cbde3010cc.png)

Your Phone:
![image](https://user-images.githubusercontent.com/1378165/108281590-e0e0d080-7134-11eb-8807-c0c0ea6477c6.png)

Word:
![image](https://user-images.githubusercontent.com/1378165/108281619-efc78300-7134-11eb-946e-b1528bd0299f.png)

## Scope
Must allow buttons, dropdowns and text/searchboxes

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow developers to accomplish commong title-bar patterns used in many other Windows Apps and familiar to users | Must |
| This proposal will allow end the developer to maximize content area by moving common features to the title bar | Must |

**Comment by mdtauk** [comment]
Can't you already do this with the Extend Into Titlebar API and SetTitleBar API?

**Comment by marck7jr** [comment]
@dotMorten I made a templated control to customize the TitleBar with interactable objects for UWP at first impression it needs two panels: one for the draggable area and non-interactive items and other one for put any kind of content rendering through ContentPresenter control and I just need to manipulate they Z-Index. I'll try here and report the results, just commenting if you want to try too.

**Comment by dotMorten** [comment]
@mdtauk No. As mentioned above they don't receive input. Looks like once you call SetTitleBar, all mouse input is diverted to dragging the window around.

@marck7jr I should point out this if for Win32 not UWP

**Comment by marck7jr** [comment]
@dotMorten I tried to "recreate" the templated control that I am using in another [project](https://www.microsoft.com/en-us/p/visum-photo-viewer/9n1x3z50blm8?activetab=pivot:overviewtab) and it "works" but events like Click didn't fire, but the curious is events like PointerEntered works fine (but is a little tricker to force the button to get the "Hover" state style). I made a little simple [project](https://github.com/marck7jr/winui-titlebarapi-sample) with the templated control sample, it can help to test this API in the future.

WinUI Desktop:
![image](https://user-images.githubusercontent.com/30981574/108288451-fcfb6680-7162-11eb-9a26-c3fefc01319f.png)

UWP:
![image](https://user-images.githubusercontent.com/30981574/108288904-cd009300-7163-11eb-9a8f-5bc72d1d7096.png)

**Comment by marb2000** [comment]
Thanks @dotMorten for opening the proposal. Thanks @marck7jr for the sample code. 

**Possible Solutions**
1. Extend the Window object with a property which is a collection of draggable areas, no just one. Something like SetTitleBars.Add(UIElement).
2. UIElements has a property to get the input regardless it's on the TitleBar (doable?)

**Comment by mdtauk** [comment]
> Thanks @dotMorten for opening the proposal. Thanks @marck7jr for the sample code.
> 
> **Possible Solutions**
> 
> 1. Extend the Window object with a property which is a collection of draggable areas, no just one. Something like SetTitleBars.Add(UIElement).
> 2. UIElements has a property to get the input regardless it's on the TitleBar (doable?)

Can you not allow setting the panel as the Titlebar, and use a Transparent brush to allow for grabbing.  Then any controls within that panel are rendered above, and so accept hit targets?  This is how UWP does it.

**Comment by bkaankose** [comment]
UWP handles this very well. If this is going to be implemented better to keep the same structure for consistency.

**Comment by LadderLogic** [comment]
@marb2000, @pratikone  any updates on this feature?

**Comment by softworkz** [comment]
Isn't this implemented already?

See [InputNonClientPointerSource.SetRegionRects Method](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.input.inputnonclientpointersource.setregionrects?view=windows-app-sdk-1.6)

I suppose this one can be closed.

**Comment by iablaauw-MS** [comment]
Yes this feature is complete! 🎉 

When using [ExtendsContentIntoTitlebar](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window.extendscontentintotitlebar?view=windows-app-sdk-1.6#microsoft-ui-xaml-window-extendscontentintotitlebar), you can also use [InputNonClientPointerSource.SetRegionRects Method](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.input.inputnonclientpointersource.setregionrects?view=windows-app-sdk-1.6) with a region kind of `Passthrough` to setup rects so that controls in the titlebar will get input.

This is admittedly a fair amount of manual work to get right, so there is upcoming work on a full `TitleBar` control that makes this even easier.
